src/traces/scatterternary/calc.js

```diff
@@ -0,0 +1,97 @@
+/**
+* Copyright 2012-2016, Plotly, Inc.
+* All rights reserved.
+*
+* This source code is licensed under the MIT license found in the
+* LICENSE file in the root directory of this source tree.
+*/
+
+
+'use strict';
+
+var isNumeric = require('fast-isnumeric');
+
+var Axes = require('../../plots/cartesian/axes');
+var Lib = require('../../lib');
+
+var subTypes = require('../scatter/subtypes');
+var calcMarkerColorscale = require('../scatter/marker_colorscale_calc');
+
+var dataArrays = ['a', 'b', 'c'];
+var arraysToFill = {a: ['b', 'c'], b: ['a', 'c'], c: ['a', 'b']};
+
+
+module.exports = function calc(gd, trace) {
+    var ternary = gd._fullLayout[trace.subplot],
+        displaySum = ternary.sum,
+        normSum = trace.sum || displaySum;
+
+    var i, j, dataArray, newArray, fillArray1, fillArray2;
+
+    // fill in one missing component
+    for(i = 0; i < dataArrays.length; i++) {
+        dataArray = dataArrays[i];
+        if(trace[dataArray] && trace[dataArray].length) continue;
+
+        fillArray1 = trace[arraysToFill[dataArray][0]];
+        fillArray2 = trace[arraysToFill[dataArray][1]];
+        newArray = new Array(fillArray1.length);
+        for(j = 0; j < fillArray1.length; j++) {
+            newArray[j] = normSum - fillArray1[j] - fillArray2[j];
+        }
+        trace[dataArray] = newArray;
+    }
+
+    // make the calcdata array
+    var serieslen = trace.a.length;
+    var cd = new Array(serieslen);
+    var a, b, c, norm, x, y;
+    for(i = 0; i < serieslen; i++) {
+        a = trace.a[i];
+        b = trace.b[i];
+        c = trace.c[i];
+        if(isNumeric(a) && isNumeric(b) && isNumeric(c)) {
+            a = Number(a);
```

Comment:
![image](https://cloud.githubusercontent.com/assets/6675409/14254822/6d45a25a-fa5f-11e5-91f9-e472ef346f08.png)

reference (because jsPerf has unreliable lately):
https://github.com/etpinard/plotly.js-benchmarks/blob/master/suites/coerce-to-number.js